File: Leetcode/1765. Map of Highest Peak.py
class Solution:
    def highestPeak(self, isWater: List[List[int]]) -> List[List[int]]:
        res=[[-1]*len(isWater[0]) for _ in range(len(isWater))]
        q=deque()
        for i in range(len(isWater)):
            for j in range(len(isWater[0])):
                if isWater[i][j]==1:
                    q.append([i,j])
                    res[i][j]=0
        dx=[-1,0,1,0]
        dy=[0,-1,0,1]
        while q:
            i,j=q[0]
            q.popleft()
            for u in range(4):
                nx=i+dx[u]
                ny=j+dy[u]
                if nx>=0 and nx<len(isWater) and ny<len(isWater[0]) and ny>=0 and res[nx][ny]==-1:
                    q.append([nx,ny])
                    res[nx][ny]=res[i][j]+1
        return res

        

File: Leetcode/1765. Map of Highest Peak.cpp
class Solution {
public:

    bool isValid(int i,int j,int m,int n){
        return i>=0 && j>=0 && j<n && i<m;
    }
    vector<vector<int>> highestPeak(vector<vector<int>>& isWater) {
        int m=isWater.size();
        int n=isWater[0].size();
        vector<int> dx={-1,0,1,0};
        vector<int> dy={0,-1,0,1};
        vector<vector<int>> res(m,vector<int>(n,-1));
        queue<pair<int,int>> q;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(isWater[i][j]==1){
                    q.push(make_pair(i,j));
                    res[i][j] = 0;
                }
            }
        }
        while(!q.empty()){
            auto [i,j]=q.front();
            q.pop();
            for (int k=0;k<4;k++) {
                int ni=i+dx[k];
                int nj=j+dy[k];
                if (isValid(ni, nj, m, n) && res[ni][nj] == -1) {
                    res[ni][nj] = res[i][j] + 1; 
                    q.push({ni, nj});
                }
            }
        }
        
        return res;
    }
};
